### PR TITLE
[SPARK-9612][ML] Add instance weight support for GBTs

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
add weight support for GBTs by sampling data before passing it to trees and then passing weights to trees

in summary:
1, add setters of `minWeightFractionPerNode` & `weightCol`
2, update input types in private methods from `RDD[LabeledPoint]` to `RDD[Instance]`:
`DecisionTreeRegressor.train`, `GradientBoostedTrees.run`, `GradientBoostedTrees.runWithValidation`, `GradientBoostedTrees.computeInitialPredictionAndError`, `GradientBoostedTrees.computeError`, 
`GradientBoostedTrees.evaluateEachIteration`, `GradientBoostedTrees.boost`, `GradientBoostedTrees.updatePredictionError`
3, add new private method `GradientBoostedTrees.computeError(data, predError)` to compute average error, since original `predError.values.mean()` do not take weights into account.
4, add new tests

### Why are the changes needed?
GBTs should support sample weights like other algs

### Does this PR introduce any user-facing change?
yes, new setters are added

### How was this patch tested?
existing & added testsuites

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Wow, @zhengruifeng . This is really a long standing JIRA. 👍

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala [hunk 1]
```diff
@@ -296,6 +304,35 @@ class GBTRegressorSuite extends MLTest with DefaultReadWriteTest {
...
+    for ((maxIter, maxDepth) <- testParams) {
+      val estimator = new GBTRegressor()
+        .setMaxIter(maxIter)
+        .setMaxDepth(maxDepth)
+        .setSeed(seed)
+        .setMinWeightFractionPerNode(0.1)
+
+      MLTestingUtils.testArbitrarilyScaledWeights[GBTRegressionModel,
+        GBTRegressor](df.as[LabeledPoint], estimator,
+        MLTestingUtils.modelPredictionEquals(df, _ ~= _ relTol 0.1, 0.95))
```
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: Compared to `DecisionTreeRegressorSuite`, I need to limit the number of trees and loose the tolerance eps(0.99 -> 0.95) to pass the cases.
I wonder if it is due to accumulated errors among trees.

**Comment by zhengruifeng** [comment]
cc @srowen @imatiach-msft

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 2]
```diff
@@ -152,32 +155,27 @@ class GBTClassifier @Since("1.4.0") (
     set(validationIndicatorCol, value)
   }
 
+  /**
+   * Sets the value of param [[weightCol]].
+   * If this is not set or empty, we treat all instance weights as 1.0.
+   * Default is not set, so all instances have weight one.
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: nit: update comment to 
```
By default the weightCol is not set, so all instances have weight 1.0.
```

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 3]
```diff
@@ -152,32 +155,27 @@ class GBTClassifier @Since("1.4.0") (
...
       dataset: Dataset[_]): GBTClassificationModel = instrumented { instr =>
-    val categoricalFeatures: Map[Int, Int] =
-      MetadataUtils.getCategoricalFeatures(dataset.schema($(featuresCol)))
+    val categoricalFeatures = MetadataUtils.getCategoricalFeatures(dataset.schema($(featuresCol)))
 
     val withValidation = isDefined(validationIndicatorCol) && $(validationIndicatorCol).nonEmpty
 
-    // We copy and modify this from Classifier.extractLabeledPoints since GBT only supports
-    // 2 classes now.  This lets us provide a more precise error message.
-    val convert2LabeledPoint = (dataset: Dataset[_]) => {
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: the error message here was much nicer:
```
GBTClassifier currently only supports binary classification.
```
than the new one in extractInstances.  Perhaps it would be nicer to keep this custom error message, or pass some part of the message to the extractInstances method.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala [hunk 4]
```diff
@@ -151,28 +154,35 @@ class GBTRegressor @Since("1.4.0") (@Since("1.4.0") override val uid: String)
     set(validationIndicatorCol, value)
   }
 
+  /**
+   * Sets the value of param [[weightCol]].
+   * If this is not set or empty, we treat all instance weights as 1.0.
+   * Default is not set, so all instances have weight one.
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: nit: update comment, same as above

File: mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala [hunk 5]
```diff
@@ -68,7 +68,7 @@ class GradientBoostedTrees private[spark] (
   def run(input: RDD[LabeledPoint]): GradientBoostedTreesModel = {
     val algo = boostingStrategy.treeStrategy.algo
     val (trees, treeWeights) = NewGBT.run(input.map { point =>
-      NewLabeledPoint(point.label, point.features.asML)
+      NewLabeledPoint(point.label, point.features.asML).toInstance
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala]: can we create an instance directly from label, features - seems a bit too much to create a temporary LabeledPoint object that's unused?

File: mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala (hunk 1 above)
Comment (imatiach-msft) [reply mllib/src/test/scala/org/apache/spark/ml/regression/GBTRegressorSuite.scala]: interesting, will need to take a closer look...

**Review by imatiach-msft — COMMENTED** [review]
added some initial comments, will need to take a closer look into how the weights are used in the boosting more

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 6]
```diff
@@ -152,32 +155,27 @@ class GBTClassifier @Since("1.4.0") (
...
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+
   override protected def train(
       dataset: Dataset[_]): GBTClassificationModel = instrumented { instr =>
-    val categoricalFeatures: Map[Int, Int] =
-      MetadataUtils.getCategoricalFeatures(dataset.schema($(featuresCol)))
+    val categoricalFeatures = MetadataUtils.getCategoricalFeatures(dataset.schema($(featuresCol)))
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: nit: can this line be moved above where it is used:

val boostingStrategy = super.getOldBoostingStrategy(categoricalFeatures, OldAlgo.Classification)

Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: probably good to move boostingStrategy below as well

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 7]
```diff
@@ -106,13 +106,13 @@ private[spark] object GradientBoostedTrees extends Logging {
...
+      data: RDD[Instance],
       initTreeWeight: Double,
       initTree: DecisionTreeRegressionModel,
       loss: OldLoss): RDD[(Double, Double)] = {
-    data.map { lp =>
-      val pred = updatePrediction(lp.features, 0.0, initTree, initTreeWeight)
-      val error = loss.computeError(pred, lp.label)
+    data.map { case Instance(label, _, features) =>
+      val pred = updatePrediction(features, 0.0, initTree, initTreeWeight)
+      val error = loss.computeError(pred, label)
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: hmm shouldn't the loss be weighted by the weight column value here?  seems a bit strange to ignore the weight column here

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 8]
```diff
@@ -129,20 +129,17 @@ private[spark] object GradientBoostedTrees extends Logging {
...
       treeWeight: Double,
       tree: DecisionTreeRegressionModel,
       loss: OldLoss): RDD[(Double, Double)] = {
-
-    val newPredError = data.zip(predictionAndError).mapPartitions { iter =>
-      iter.map { case (lp, (pred, error)) =>
-        val newPred = updatePrediction(lp.features, pred, tree, treeWeight)
-        val newError = loss.computeError(newPred, lp.label)
+    data.zip(predictionAndError).map {
+      case (Instance(label, _, features), (pred, _)) =>
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: same thing here - it seems like we are ignoring the weight column but intuitively it seems like it should be included, could you explain the reasoning behind it being excluded here?

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 9]
```diff
@@ -166,29 +163,50 @@ private[spark] object GradientBoostedTrees extends Logging {
...
-    data.map { lp =>
+    val (errSum, weightSum) = data.map { case Instance(label, weight, features) =>
       val predicted = trees.zip(treeWeights).foldLeft(0.0) { case (acc, (model, weight)) =>
-        updatePrediction(lp.features, acc, model, weight)
+        updatePrediction(features, acc, model, weight)
       }
-      loss.computeError(predicted, lp.label)
-    }.mean()
+      (loss.computeError(predicted, label) * weight, weight)
+    }.treeReduce{ case ((err1, weight1), (err2, weight2)) =>
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: nit: spacing of { after treeReduce

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 10]
```diff
@@ -197,41 +215,41 @@ private[spark] object GradientBoostedTrees extends Logging {
...
-    val evaluation = remappedData.map { point =>
-      treesIndices.map { idx =>
-        val prediction = broadcastTrees.value(idx)
-          .rootNode
-          .predictImpl(point.features)
-          .prediction
-        prediction * localTreeWeights(idx)
+    val numTrees = trees.length
+
+    val (errSum, weightSum) = remappedData.mapPartitions { iter =>
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: trying to understand this code - why are the trees broadcast here but the treeWeights are not?

Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: just to be clear the previous code is doing this as well, I just don't understand why the treeWeights aren't broadcast either

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 7 above)
Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: oh, reading some of the other code this looks like unweighted error.  That seems very confusing.  I think we could improve this code structure a bit more.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 8 above)
Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: oh, reading some of the other code this looks like unweighted error.  That seems very confusing.  I think we could improve this code structure a bit more.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 7 above)
Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: what would be the problem with this returning weighted error and getting rid of the computeError function?

File: mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala (hunk 5 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala]: Yes, it is better to directly create `Instance`

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 11]
```diff
@@ -299,26 +317,25 @@ private[spark] object GradientBoostedTrees extends Logging {
     baseLearners(0) = firstTreeModel
     baseLearnerWeights(0) = firstTreeWeight
 
-    var predError: RDD[(Double, Double)] =
-      computeInitialPredictionAndError(input, firstTreeWeight, firstTreeModel, loss)
+    var predError = computeInitialPredictionAndError(input, firstTreeWeight, firstTreeModel, loss)
     predErrorCheckpointer.update(predError)
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: it would be nice if we could checkpoint the weighted instead of unweighted prediction error, which ties into the earlier comment on why methods like computeInitialPredictionAndError can't return the weighted prediction error

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 10 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: in this place, I just followed preivous impl. I am neutral on it.

**Comment by zhengruifeng** [comment]
@imatiach-msft Thanks for reviewing!
As to the points on *weighted* prediction error:
After previous discussions, we should sample the data without weights, and pass the weights into the base model (decision tree).
So the input passed to a decsion tree, should contain the label (unweighted prediction error) and the instance weights (which will be also used in `minWeightFractionPerNode`). In this way, I guess we do not need to cache weighted error.
Moreover, the code `predError.values.mean()` with weighted `predError` is not equal to the average weighted error in this PR.

PS: If I recall correctly, XGBoost pass weighted gradients and hessions into base learner. It use minimum hession (`min_child_weight`) to limit tree growth, which is quite different from MLLIB.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 10 above)
Comment (imatiach-msft) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: sounds good

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala [hunk 12]
```diff
@@ -166,29 +163,50 @@ private[spark] object GradientBoostedTrees extends Logging {
...
+    errSum / weightSum
+  }
+
+  /**
+   * Method to calculate error of the base learner for the gradient boosting calculation.
+   * @param data Training dataset: RDD of `Instance`.
+   * @param predError Prediction and error.
+   * @return Measure of model error on data
+   */
+  def computeError(
```
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: maybe change the name to computeWeightedError to make that clear, since the above methods are also computing error but unweighted

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala (hunk 11 above)
Comment (imatiach-msft) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala]: if we are going to keep the checkpointing of unweighted error as opposed to weighted error, then it would be nice to specify that in the name of the checkpointer:
predUnweigtedErrorCheckpointer
or alternatively add a comment to make that clear:
```
    // Note: this is checkpointing the unweighted error
    predErrorCheckpointer.update(predError) 
```

**Comment by imatiach-msft** [comment]
@zhengruifeng I see, it seems a bit confusing to have a lot of references to error as both weighted and unweighted inside the same function - for example I would prefer to only checkpoint weighted error, and computeError function doesn't suggest anything in the name about it specifically making the error weighted - but as long as we have good documentation and variable names in the code to help distinguish which variable is for what I think it should be fine

**Review by imatiach-msft — APPROVED** [review]
great feature!  LGTM!

**Comment by zhengruifeng** [comment]
@imatiach-msft Thanks for reviewing and your previous works on decision tree supporting sample weights!

**Comment by zhengruifeng** [comment]
I manually tested this PR in repl in the past days, with some datasets in `/data/mllib`, set relative params to in normal ranges (for example `weight` in [1.0, 10.0], not extreme values (0.01, 1000) in the testsuits), and the results looked fine.

I think the error is accumulated among `DecisionTree`s, so I need to loose the tolerance in test suites (compared with `DecisionTreeSuites`).

I will merge this PR this week if no more comments.

**Comment by zhengruifeng** [comment]
Merged to master, thanks @imatiach-msft @srowen for reviewing!
